[clang] Implement P2582R1: CTAD from inherited constructors

antangelo opened the pull request:
There are two known issues with this initial implementation:
1. Deduction guides are not generated for explicit guides of the base defined after the initial usage of the derived class. This is a caused by a similar issue in the alias template deduction guide implementation.
2. It is currently unable to detect bases whose constructors are inherited by a using declaration with a dependent name, such as `using Derived::Base::Base;`. These using declarations in general are unresolved, so it is difficult to look them up and match them to the corresponding base before instantiation.

I intend to target clang 20 for merging this patch.

Closes #92606

antangelo committed:
[clang] Implement P2582R1: CTAD from inherited constructors

shafik commented on clang/lib/AST/DeclCXX.cpp:
```diff
@@ -2157,17 +2157,19 @@ CXXDeductionGuideDecl *CXXDeductionGuideDecl::Create(
       C, nullptr, SourceLocation(), ExplicitSpecifier(), DeclarationNameInfo(),
       QualType(), nullptr, SourceLocation(), nullptr,
-      DeductionCandidate::Normal);
+      DeductionCandidate::Normal, nullptr, false);
```
```suggestion
      DeductionCandidate::Normal, /*SourceDG=*/nullptr, /*IsGeneratedFromInheritedConstructor =*/false);
```

This is based on [bugprone argument comment](https://clang.llvm.org/extra/clang-tidy/checks/bugprone/argument-comment.html) convention.

antangelo replied:
Thanks, I've updated those arguments as you suggested. Do you want me to also update the others, or just those two new ones?

shafik reviewed (commented):
Quick drive by comment

antangelo committed:
Fix PR comment

antangelo committed:
Merge remote-tracking branch 'upstream/main' into p2582r1

antangelo committed:
clang-format

hokein commented:
Thanks for implementing it! 

I haven't looked into the implementation details yet.
Before doing so, I want to ensure I understand the standard correctly (the standard's wording is a bit hard to follow).

Per C++ [over.match.class.deduct#1.10](https://eel.is/c++draft/over.match.class.deduct#1.10):

> In addition, if C is defined and inherits constructors ([[namespace.udecl]](https://eel.is/c++draft/namespace.udecl)) from a direct base class denoted in the [base-specifier-list](https://eel.is/c++draft/class.derived.general#nt:base-specifier-list) by a [class-or-decltype](https://eel.is/c++draft/class.derived.general#nt:class-or-decltype) B, let A be an alias template whose template parameter list is that of C and whose [defining-type-id](https://eel.is/c++draft/dcl.name#nt:defining-type-id) is B[.](https://eel.is/c++draft/over.match.class.deduct#1.sentence-7) If A is a deducible template ([[dcl.type.simple]](https://eel.is/c++draft/dcl.type.simple)), the set contains the guides of A with the return type R of each guide replaced with typename CC<R>​::​type given a class template
template <typename> class CC;
whose primary template is not defined and with a single partial specialization whose template parameter list is that of A and whose template argument list is a specialization of A with the template argument list of A ([[temp.dep.type]](https://eel.is/c++draft/temp.dep.type)) having a member typedef type designating a template specialization with the template argument list of A but with C as the template[.](https://eel.is/c++draft/over.match.class.deduct#1.sentence-8)

Let's use the following example to illustrate:

```cpp
template <typename T1> struct B {
  B(T1);
};
template <typename T2> struct C : public B<T2> {
  using B<T2>::B;
};
```

Here, `C` is the struct `C`, and `B` is `B<T2>`.


### Steps:

1. **Find the inherited constructor for class `C`:** `B(T1)`.

2. **Synthesize an alias template `A`:** `template<typename T2> using A = B<T2>;`

3. **Check if `A` is a deducible template per https://eel.is/c++draft/dcl.type.simple#3:** here A is deducible

4. **Synthesize the deduction guide for `A`:** the one we have interests is `auto (T2) -> B(T2)`

5.  **Replace the return type with `CC<B<T2>>::type`:**

    ```cpp
    template <typename> class CC;

    template <typename T2>
    class CC<B<T2>> {
    public:
        typedef C<T2> type;
    };
    ```

The we get the final deduction guide for C: `auto (T2) -> CC<B<T2>>::type`;

antangelo commented:
> Thanks for implementing it!
> 
> I haven't looked into the implementation details yet. Before doing so, I want to ensure I understand the standard correctly (the standard's wording is a bit hard to follow).
> 

That looks correct to my understanding of the standard.

mizvekov commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -1254,8 +1341,221 @@ void DeclareImplicitDeductionGuidesForTypeAlias(
+
+  // ... given a class template `template <typename> class CC;`
+  // whose primary template is not defined ...
+  auto *TParam = TemplateTypeParmDecl::Create(
+      Context, DC, SourceLocation(), SourceLocation(),
+      Template->getTemplateDepth(), 0, nullptr, false, false);
+  auto *MapperTPL = TemplateParameterList::Create(
+      Context, SourceLocation(), SourceLocation(),
+      ArrayRef<NamedDecl *>(TParam), SourceLocation(), nullptr);
```
These should be implicit as well.

antangelo replied:
I've made the TemplateTypeParmDecl implicit, but I don't see any way of marking the parameter list as implicit (unless you were only referring to the parameter).

mizvekov replied:
Yeah I was only referring to the parameters.
Only declarations can be implicit, which a parameter list is not.

mizvekov reviewed (commented):
Just a couple of nits, I will do a more comprehensive review later.

antangelo committed:
Merge remote-tracking branch 'upstream/main' into p2582r1

antangelo committed:
Remove QualType reference and set mapper template param to implicit

antangelo commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -980,19 +986,69 @@ getRHSTemplateDeclAndArgs(Sema &SemaRef, TypeAliasTemplateDecl *AliasTemplate) {
+                        DeducingTemplate->getBeginLoc(), DeclarationName());
+  if (!ReturnTypeTSI || Trap.hasErrorOccurred())
+    return {nullptr, QualType()};
+  const QualType &ReturnType = ReturnTypeTSI->getType();
```
Thanks, replaced this with a copy.

hokein commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -936,19 +940,21 @@ Expr *buildIsDeducibleConstraint(Sema &SemaRef,
       Context.getTrivialTypeSourceInfo(
           Context.getDeducedTemplateSpecializationType(
-              TemplateName(AliasTemplate), /*DeducedType=*/QualType(),
+              TemplateName(TD), /*DeducedType=*/QualType(),
```
```cpp
template <typename T> struct B {
  B(T);
};
template <typename T> struct C : public B<T> {
  using B<T>::B;
};

C c(1);

// implicit code
// template <typename> class CC;
//
// template <typename T>
// class CC<B<T>> {
// public:
//    typedef C<T> type;
// };
```
If I read the code correctly,  we will build the `__is_deducible(C, typename CC<B<T>>::type)` for the deduction guide of `C` (`auto (T) -> typename CC<B<T>>::type`).  I'm not sure this is correct.

Intuitively, this constraint is not necessary, if `CC<B<T>>::type` is valid, it must be the type of `C<T>`, then `is_deducible(C, C<T>)` is always true.

Unfortunately, the standard doesn't specify the transformation of the associated constraint (oversight?). If we only replace the return type, then the constraint will be `is_deducible(B,  typename CC<B<T>>::type)`  which is always false. 

Perhaps we should always drop the `is_deducible` constraint for this case.

antangelo replied:
I think you're right that the `is_deducible` constraint is technically unnecessary-- as far as I can tell, any case where this constraint would fail would cause the `CC<R>::type` instantiation to fail before the constraint is evaluated.
There are some other caveats with removing it, however. Consider this example:

```c++
template<typename T>
struct Base1 { };

template<typename T>
struct Base2 { };

template<typename T = int>
struct Derived : public Base1<T>, Base2<T> {
    using Base1<T>::Base1;
    using Base2<T>::Base2;
};

Derived d;
```
https://godbolt.org/z/4YcdvTMov

Under [[over.match.best.general]p2.5](https://eel.is/c++draft/over.match.best.general#2.5) (and more specifically [[temp.func.order]p6.4](https://eel.is/c++draft/temp.func.order#6.4)), the bases' deduction guides are more specialized than any from `Derived` (due to the `is_deducible` constraint), and so the implementation as it is sees this as ambiguous.

Since the standard doesn't specify what to do with the deducible constraint, and unless there's an oversight here, I assume this is what it wants to happen.

Given that [[over.match.best.general]p2.10](https://eel.is/c++draft/over.match.best.general#2.10) would otherwise give priority to the `Derived` deduction guide, it's possible there's an oversight, but I'm not sure if that warrants removing the deducible constraint yet.

cor3ntin replied:
Per https://eel.is/c++draft/over.match.class.deduct#1.sentence-7, 
I think we imply we should copy the constraints of the made up alias. WDYT?

Feel free to open an issue here though https://github.com/cplusplus/CWG/issues

antangelo replied:
Based on the current writing of the standard I think we should be copying all of the constraints over, including the deducible constraint. However, it's probably not the intent to break existing code as in the above example.

I have filed https://github.com/cplusplus/CWG/issues/607 . Is the best course of action in the meantime to update the implementation in this PR to accept that above code, or leave it as breaking until CWG responds?

antangelo replied:
I believe I forgot to acknowledge this comment. I am skipping the construction of the deducible constraint as of [0245459](https://github.com/llvm/llvm-project/pull/98788/commits/024545921b5ac4c8e9fa4781b492cf719877cb05)

hokein commented on clang/test/SemaCXX/cxx23-ctad-inherited-ctors.cpp:
```diff
@@ -0,0 +1,260 @@
+// RUN: %clang_cc1 -fsyntax-only -std=c++23 -verify %s
+
```
Can you add testcases for the template parameter packs? e.g. 

```
template <typename ...T> struct B {
  B(T...);
};
template <typename ...T> struct C : public B<T...> {
  using B<T...>::B;
};

C c(1);
```

antangelo replied:
Thanks for catching this, I've fixed this case and added a test

hokein commented on clang/test/SemaCXX/cxx23-ctad-inherited-ctors.cpp:
```diff
@@ -0,0 +1,260 @@
+// RUN: %clang_cc1 -fsyntax-only -std=c++23 -verify %s
```
nit: you need to adjust the diagnostic messages after the https://github.com/llvm/llvm-project/commit/c7bfc41860a6abe5c92dc5afb47348b0c9e69963

antangelo replied:
Thanks, fixed

hokein reviewed (commented):
I haven't finished the review yet, some initial comments.

antangelo committed:
Merge remote-tracking branch 'upstream/main' into p2582r1

antangelo committed:
Update test notes to account for canonical template arg changes

antangelo committed:
Rely on Using decls instead of bases for determining inherited constructors

antangelo committed:
Use static_assert(__is_same(...)) to verify types in tests

antangelo committed:
Update comment for BuildDeductionGuideForTypeAlias and fix nits

antangelo committed:
Use enum to specify deduction guide source kind

antangelo committed:
Fix parameter pack case

antangelo commented on clang/include/clang/AST/DeclCXX.h:
```diff
@@ -1957,17 +1957,26 @@ class CXXDeductionGuideDecl : public FunctionDecl {
+  // The deduction guide, if any, that this deduction guide was generated from,
+  // in the case of alias template deduction or CTAD from inherited
+  // constructors. The bool member indicates whether this deduction guide is
+  // generated from an inherited constructor.
```
Thanks for the suggestion, I've updated it to use an enum here.

antangelo commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -980,19 +986,69 @@ getRHSTemplateDeclAndArgs(Sema &SemaRef, TypeAliasTemplateDecl *AliasTemplate) {
     return nullptr;
 
+  if (!DeducingTemplate)
+    DeducingTemplate = AliasTemplate;
```
Done

antangelo commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -980,19 +986,69 @@ getRHSTemplateDeclAndArgs(Sema &SemaRef, TypeAliasTemplateDecl *AliasTemplate) {
-BuildDeductionGuideForTypeAlias(Sema &SemaRef,
-                                TypeAliasTemplateDecl *AliasTemplate,
-                                FunctionTemplateDecl *F, SourceLocation Loc) {
+FunctionTemplateDecl *BuildDeductionGuideForTypeAlias(
```
Done, updated to add mention of the new parameters for the inherited constructors case

antangelo committed:
Merge remote-tracking branch 'upstream/main' into p2582r1

antangelo committed:
clang-format

antangelo committed:
Merge branch 'main' into p2582r1

antangelo commented:
Friendly ping

hokein reviewed (commented):
The current implementation strictly follows the standard's specification for replacing the return type of an alias deduction guide.

However, this approach adds some implicit declarations in the AST (in terms of the partial class template specialization), which is cumbersome. There is an alternative approach, which is to perform an earlier deduction (when building the deduction guide) instead of instantiation time deduction, it would allow us to avoid having these implicit decls in the AST. This is the approach taken by gcc.

The downside is that we diverge from the standard, which might result in subtle differences. What do people think? @cor3ntin @mizvekov @zygoloid

antangelo commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -1216,10 +1308,225 @@ void DeclareImplicitDeductionGuidesForTypeAlias(
+    return;
+
+  // Subsitute any parameters with default arguments not present in the base,
+  // since partial specializations cannot have default parameters
```
Consider this example:

```c++
template<class T>
struct Base {
    Base(T t);
};

template<class InBase = int, class NotInBase = int>
struct A : public Base<InBase> {
    using Base<InBase>::Base;
};

A a(10);
```
(accepted by gcc: https://godbolt.org/z/jbj1WMq9d)

To my understanding of the standard, we should generate:
```c++
template<class InBase = int, class NotInBase = int>
using Alias = Base<InBase>;

template<typename T> class CC;
template<typename InBase, typename NotInBase>
class CC<Base<InBase>> {
    typedef A<InBase, NotInBase> type;
};
```
along with the transformed alias deduction guides. With these in place, that deduction will fail, because it is not possible to instantiate the partial specialization of `CC<R>` for the deduction guide return type `R` (which is `Base<InBase>` in practice).

The reason this substitution fails is because `NotInBase` cannot be determined from `Base<InBase>`. Since partial specializations cannot have default arguments, the fact that `NotInBase` is defaulted in the derived class is not respected.

The solution I came up with for this problem is to rewrite the template parameters to substitute in these template arguments which have defaults in the derived class, but do not appear in the base. Then, the partial specialization member typedef can refer to them directly. In practice, these arguments cannot be deduced to anything else anyway.

This issue does not exist for defaulted parameters that do appear in the base-- those will substitute fine into the partial specialization. I also opted not to bail in the case where there is a template parameter with no default that does not appear in the base; the template is not deducible in this case, but I opt to generate a deduction guide anyway so we can emit a diagnostic.

The TPL does indeed still contain the default arguments, but clang's implementation as it is will ignore them. I agree that we probably shouldn't rely on this, though.

antangelo replied:
I don't believe it simplifies things significantly, we still need to run the template instantiator to clone the parameter list, so the only additional work is checking for default arguments and substituting subsequent parameters accordingly. In addition, gcc tends to accept these test cases as valid code.
I will file another CWG issue to cover these cases.

hokein replied:
Fair enough. It would be great if you can add the filed issue in a comment.

antangelo replied:
I have added a link to the github issue for this and the other CWG issue. If CWG assigns a number to those issues I'll update the comment to use the proper number instead.

antangelo commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -1216,10 +1308,225 @@ void DeclareImplicitDeductionGuidesForTypeAlias(
+
+  // [over.match.class.deduct]p1.10
+  // Let A be an alias template whose template parameter list is that of
+  // [Template] and whose defining-type-id is [BaseTSI] ...
```
I think it is possible to get an `A` that is not deducible, here's an example similar to the above:
```c++
template<class T>
struct Base {
    Base(T t);
};

template<class InBase, class NotInBase>
struct A : public Base<InBase> {
    using Base<InBase>::Base;
};

A a(10);
```

`NotInBase` cannot be deduced from `Base`'s constructors, and the `A` that we construct will not be deducible.

However, I think even in this case it's worth emitting a deduction guide that will never succeed so we can emit a diagnostic if the user tries to use it.

antangelo replied:
I see, I had this confused with the deducible definition for the alias deduction guide constraint. According to `[dcl.type.simple]p2` the definition is recursive on `simple-template-id` being deducible. So, I think we do need to explicitly check that the base template is deducible (i.e. in cases where the base is itself an alias). Thanks for catching that!

antangelo replied:
I've added a couple test cases to the `test10` namespace where the base classes are not deducible according to [dcl.type.simple]p3 (using bases that are alias templates that do not name class templates).

antangelo commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -1216,10 +1308,225 @@ void DeclareImplicitDeductionGuidesForTypeAlias(
+
+  auto *MapperSpecialization = ClassTemplatePartialSpecializationDecl::Create(
+      Context, ClassTemplatePartialSpecializationDecl::TagKind::Struct, DC,
+      SourceLocation(), SourceLocation(), PartialSpecTPL, MapperTD, TAL,
```
Are you referring to the parameters themselves being incorrect between the alias/partial specialization, or the reuse of the same TemplateParameterList instance for the two decls being incorrect?
If it's the latter, we reuse the class template's TPL when creating implicit deduction guides, is that also incorrect?

antangelo replied:
Done, I am now cloning the alias TPL and stripping default arguments from it to create a new TPL for the partial specialization.

antangelo committed:
Merge remote-tracking branch 'upstream/main' into p2582r1

antangelo committed:
Address PR feedback

- Get ReturnType from FPT TSI instead of returning it in pair
- Rename DerivedClassMapperType to CCType
- Rename DeducingTemplate to DeducedTemplate to match
  CXXDeductionGuideDecl member function `getDeducedTemplate()`
- Address nits
- Move FIXME for generating inherited ctor dguides declared after the
  first time DeclareImplicitDeductionGuides runs to the
  hasDeclaredDeductionGuides check
- Move inherited constructor relater parameters to the alias deduction
  guide functions into an optional struct

antangelo committed:
Run clang-format

antangelo commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -944,12 +950,63 @@ getRHSTemplateDeclAndArgs(Sema &SemaRef, TypeAliasTemplateDecl *AliasTemplate) {
+// [over.match.class.deduct]p1.10:
+// ... the set contains the guides of A with the return type R
+// of each guide replaced with `typename CC<R>::type` ...
+std::pair<TypeSourceInfo *, QualType>
```
Done

antangelo commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -944,12 +950,63 @@ getRHSTemplateDeclAndArgs(Sema &SemaRef, TypeAliasTemplateDecl *AliasTemplate) {
+// of each guide replaced with `typename CC<R>::type` ...
+std::pair<TypeSourceInfo *, QualType>
+buildInheritedConstructorDeductionGuideType(
+    Sema &SemaRef, TypeSourceInfo *DerivedClassMapperType,
```
Done, I renamed it to `CCType` within the `InheritedConstructorDeductionInfo` struct.

antangelo commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -944,12 +950,63 @@ getRHSTemplateDeclAndArgs(Sema &SemaRef, TypeAliasTemplateDecl *AliasTemplate) {
+std::pair<TypeSourceInfo *, QualType>
+buildInheritedConstructorDeductionGuideType(
+    Sema &SemaRef, TypeSourceInfo *DerivedClassMapperType,
+    TemplateDecl *DeducingTemplate, TypeSourceInfo *SourceGuideTSI) {
```
It is the template we are building the deduction guide for, i.e. the derived class template. I have renamed it to `DerivedClassTemplate` in the context of inherited constructors and `DeducedTemplate` otherwise (to match the naming of `CXXDeductionGuideDecl::getDeducedTemplate()`) to make it a bit more clear.

antangelo commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -1388,6 +1695,33 @@ void Sema::DeclareImplicitDeductionGuides(TemplateDecl *Template,
 
+  // FIXME: Handle explicit deduction guides from inherited constructors
+  // when the base deduction guides are declared after this has first run
+  CXXRecordDecl *TemplatedDecl = Pattern->getTemplatedDecl();
```
Done

antangelo commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -1388,6 +1695,33 @@ void Sema::DeclareImplicitDeductionGuides(TemplateDecl *Template,
   if (!AddedAny)
     Transform.buildSimpleDeductionGuide(std::nullopt);
 
+  // FIXME: Handle explicit deduction guides from inherited constructors
```
Done

antangelo commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -944,12 +950,63 @@ getRHSTemplateDeclAndArgs(Sema &SemaRef, TypeAliasTemplateDecl *AliasTemplate) {
+    Sema &SemaRef, TypeAliasTemplateDecl *AliasTemplate,
+    FunctionTemplateDecl *F, SourceLocation Loc,
+    TemplateDecl *DeducingTemplate = nullptr,
+    TypeSourceInfo *DerivedClassMapperType = nullptr) {
```
Done

hokein commented:
> The current implementation strictly follows the standard's specification for replacing the return type of an alias deduction guide.
> 
> However, this approach adds some implicit declarations in the AST (in terms of the partial class template specialization), which is cumbersome. There is an alternative approach, which is to perform an earlier deduction (when building the deduction guide) instead of instantiation time deduction, it would allow us to avoid having these implicit decls in the AST. This is the approach taken by gcc.
> 
> The downside is that we diverge from the standard, which might result in subtle differences. What do people think? @cor3ntin @mizvekov @zygoloid

I think implementation is appropriate since it follows to the standard’s specification. We should implement it correctly at first. Any further improvements or optimizations can be explored in the future if they become necessary.

antangelo committed:
Merge remote-tracking branch 'upstream/main' into p2582r1

antangelo committed:
Merge remote-tracking branch 'upstream/main' into p2582r1

antangelo committed:
Address nits, check base template deducibility, clone TPL for partial spec

antangelo committed:
clang-format

antangelo committed:
Merge remote-tracking branch 'upstream/main' into p2582r1

antangelo commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -1216,10 +1324,225 @@ void DeclareImplicitDeductionGuidesForTypeAlias(
+      ArrayRef<NamedDecl *>(TParam), SourceLocation(), nullptr);
+
+  std::string MapperDeclName =
+      (Twine("__ctad_mapper_") + BaseTD->getName() + "_to_" +
```
Done

antangelo commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -1216,10 +1324,225 @@ void DeclareImplicitDeductionGuidesForTypeAlias(
+      Context, DC, SourceLocation(), BaseLoc,
+      TransformedBase->getType().getBaseTypeIdentifier(), TransformedBase);
+  std::string AliasDeclName =
+      (Twine("__ctad_alias_") + BaseTD->getName() + "_to_" +
```
Done

antangelo commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -944,12 +950,67 @@ getRHSTemplateDeclAndArgs(Sema &SemaRef, TypeAliasTemplateDecl *AliasTemplate) {
+FunctionTemplateDecl *BuildDeductionGuideForTypeAlias(
+    Sema &SemaRef, TypeAliasTemplateDecl *AliasTemplate,
+    FunctionTemplateDecl *F, SourceLocation Loc,
+    std::optional<InheritedConstructorDeductionInfo> FromInheritedCtor =
```
Done

antangelo commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -944,12 +950,67 @@ getRHSTemplateDeclAndArgs(Sema &SemaRef, TypeAliasTemplateDecl *AliasTemplate) {
+// ... the set contains the guides of A with the return type R
+// of each guide replaced with `typename CC<R>::type` ...
+TypeSourceInfo *buildInheritedConstructorDeductionGuideType(
+    Sema &SemaRef, InheritedConstructorDeductionInfo Info,
```
Done

antangelo commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -944,12 +950,67 @@ getRHSTemplateDeclAndArgs(Sema &SemaRef, TypeAliasTemplateDecl *AliasTemplate) {
 }
 
+struct InheritedConstructorDeductionInfo {
+  TemplateDecl *DerivedClassTemplate;
```
Done

antangelo commented:
Friendly ping

antangelo committed:
Merge branch 'main' of github.com:llvm/llvm-project into p2582r1

antangelo commented:
Friendly ping

mizvekov commented on clang/lib/AST/DeclCXX.cpp:
```diff
@@ -2211,17 +2211,20 @@ CXXDeductionGuideDecl *CXXDeductionGuideDecl::Create(
       QualType(), nullptr, SourceLocation(), nullptr,
-      DeductionCandidate::Normal);
+      DeductionCandidate::Normal, /*GeneratedFrom=*/nullptr,
+      /*SK=*/SourceKind::Alias);
```
Nit: The comment is not really necessary since the spelling of the argument gives enough context.
```suggestion
      SourceKind::Alias);
```

antangelo replied:
I've removed the comment

mizvekov commented on clang/include/clang/AST/DeclCXX.h:
```diff
@@ -2007,6 +2022,21 @@ class CXXDeductionGuideDecl : public FunctionDecl {
+    SourceDeductionGuide.setPointer(DG);
+  }
+
+  SourceKind getSourceKind() const { return SourceDeductionGuide.getInt(); }
```
Since this 'kind' is meaningless in case SourceDeductionGuide is not set, should we assert here, or perhaps add a third kind, 'None'?

antangelo replied:
I've opted for adding an additional kind `None` for this case, I think it makes the code cleaner by not having `Alias` serve as the default and needing to check the pointer before being able to read the kind.

mizvekov commented on clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -10581,6 +10581,47 @@ bool clang::isBetterOverloadCandidate(
+          const auto &A1 = FPT1->getParamTypes();
+          const auto &A2 = FPT2->getParamTypes();
+          for (unsigned I = 0, N = FPT1->getNumParams(); I != N; ++I) {
+            llvm::FoldingSetNodeID ID1, ID2;
+            S.Context.getCanonicalType(A1[I]).Profile(ID1);
+            S.Context.getCanonicalType(A2[I]).Profile(ID2);
+            if (ID1 != ID2) {
+              ParamsHaveSameType = false;
+              break;
+            }
```
```suggestion
            if (!S.Context.hasSameType(A1[I]), A2[I]) {
              ParamsHaveSameType = false;
              break;
            }
```
Though you can probably do better here using `llvm::all_of`.

antangelo replied:
Thanks, I've updated this section to use similar logic to your suggestion along with `llvm::all_of`.

mizvekov commented on clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -11723,6 +11764,40 @@ static void DiagnoseBadDeduction(Sema &S, NamedDecl *Found, Decl *Templated,
+    if (auto *DG = dyn_cast<CXXDeductionGuideDecl>(Templated);
+        DG && DG->getSourceKind() ==
+                  CXXDeductionGuideDecl::SourceKind::InheritedConstructor) {
+      CXXDeductionGuideDecl *Source = DG->getSourceDeductionGuide();
+      assert(Source &&
+             "Inherited constructor deduction guides must have a source");
```
By calling `getSourceKind()`, you already assumed there is a source deduction guide, as otherwise the kind value is meaningless. See previous comment.

cor3ntin replied:
I think this is resolved

mizvekov commented on clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -11723,6 +11764,40 @@ static void DiagnoseBadDeduction(Sema &S, NamedDecl *Found, Decl *Templated,
+      CXXDeductionGuideDecl *Source = DG->getSourceDeductionGuide();
+      assert(Source &&
+             "Inherited constructor deduction guides must have a source");
+      auto DeducedRecordType =
+          QualType(cast<ClassTemplateDecl>(DG->getDeducedTemplate())
+                       ->getTemplatedDecl()
+                       ->getTypeForDecl(),
+                   0);
+      auto InheritedRecordType =
+          QualType(cast<ClassTemplateDecl>(Source->getDeducedTemplate())
+                       ->getTemplatedDecl()
+                       ->getTypeForDecl(),
+                   0);
```
Is the cast to QualType really necessary here? I thought the diagnostic engine could have Type* just fine.
In any case, the last argument should not be necessary.
```suggestion
      auto DeducedRecordType =
          QualType(cast<ClassTemplateDecl>(DG->getDeducedTemplate())
                       ->getTemplatedDecl()
                       ->getTypeForDecl());
      auto InheritedRecordType =
          QualType(cast<ClassTemplateDecl>(Source->getDeducedTemplate())
                       ->getTemplatedDecl()
                       ->getTypeForDecl());
```

antangelo replied:
I think both are necessary, the code does not compile for me if I try to pass in a `Type *` or try to construct QualType without the Quals parameter.

mizvekov commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -944,12 +950,72 @@ getRHSTemplateDeclAndArgs(Sema &SemaRef, TypeAliasTemplateDecl *AliasTemplate) {
+
+  MultiLevelTemplateArgumentList Args;
+  Args.addOuterTemplateArguments(Info.DerivedClassTemplate,
+                                 TemplateArgument(FPT->getReturnType()), false);
```
Please add a comment for the boolean argument.

antangelo replied:
Done

mizvekov commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -1186,17 +1281,34 @@ void DeclareImplicitDeductionGuidesForTypeAlias(
-      auto *Transformed = cast<FunctionDecl>(buildDeductionGuide(
-          SemaRef, AliasTemplate, /*TemplateParams=*/nullptr,
+
+      QualType ReturnType =
+          cast<FunctionProtoType>(FunctionType->getType())->getReturnType();
+      if (!FunctionType)
+        continue;
```
Is this check really necessary? Since you already performed a cast above, this can't be null.
Can you add a test case for this?

antangelo replied:
Thanks for catching this! It's not necessary, I believe it's an erroneous leftover from a prior refactor. In the current code path that function type is always defined. I've removed the check accordingly.

hokein reviewed (commented):
Sorry for the late response, I missed this earlier (it looks generally good).

mizvekov commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -1216,10 +1328,280 @@ void DeclareImplicitDeductionGuidesForTypeAlias(
   }
 }
 
+// Check if a template is deducible as per [dcl.type.simple]p3
```
Not as such, but I think  `Sema::GetNameFromUnqualifiedId` needs to be updated with these rules, see https://github.com/llvm/llvm-project/blob/915df1ae41652e2f595ce741dcd8f01878ef4e30/clang/lib/Sema/SemaDecl.cpp#L5796

antangelo replied:
Just to clarify, is the idea that it should look up the template name in the underlying type of the alias (possibly recursively), provided it names a deducible template?

antangelo committed:
Merge branch 'main' of github.com:llvm/llvm-project into p2582r1

antangelo committed:
clang-format

antangelo committed:
Fix constraint substitution during param list cloning

antangelo committed:
PR feedback

antangelo committed:
clang-format

antangelo commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -944,12 +950,72 @@ getRHSTemplateDeclAndArgs(Sema &SemaRef, TypeAliasTemplateDecl *AliasTemplate) {
-                                TypeAliasTemplateDecl *AliasTemplate,
-                                FunctionTemplateDecl *F, SourceLocation Loc) {
+// If F is synthesized from a base class (as an inherited constructor),
+// then the return type will be transformed using DerivedClassMapperType.
```
Thanks for catching that, I've updated the naming in that comment

antangelo commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -1216,10 +1328,280 @@ void DeclareImplicitDeductionGuidesForTypeAlias(
+      Context, DC, SourceLocation(), SourceLocation(),
+      Template->getTemplateDepth(), 0, nullptr, false, false);
+  TParam->setImplicit();
+  auto *MapperTPL = TemplateParameterList::Create(
```
Done, I've also added additional context in the name to specify which part of the `CC` type these nodes contribute to (e.g. `CCTemplateII`, `CCTemplateRD`, `CCPartialSpecializationType`, etc.)

antangelo commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -634,7 +634,7 @@ struct ConvertConstructorToDeductionGuideTransform {
 // Find all template parameters that appear in the given DeducedArgs.
 // Return the indices of the template parameters in the TemplateParams.
-SmallVector<unsigned> TemplateParamsReferencedInTemplateArgumentList(
+llvm::SmallSet<unsigned, 8> TemplateParamsReferencedInTemplateArgumentList(
```
Done

cor3ntin commented on clang/include/clang/AST/DeclCXX.h:
```diff
@@ -1960,24 +1960,38 @@ class ExplicitSpecifier {
   void anchor() override;
 
+public:
+  enum class SourceKind {
```
Can we get a better name / comment for this enum?

antangelo replied:
Done, renamed to `SourceDeductionGuideKind` to align with the class member and added a comment

hokein replied:
@cor3ntin, if you have any ideas for a better name, could you suggest one?

I think adding a few comments to the enum should suffice. Personally, I like `SourceKind` — it's short, clear, and works well as `CXXDeductionGuideDecl::SourceKind::XXX`.

On the other hand, `SourceDeductionGuideKind` feels a bit lengthy and redundant with "DeductionGuide" already in the context.

cor3ntin commented on clang/include/clang/Sema/SemaInternal.h:
```diff
@@ -58,7 +58,7 @@ inline InheritableAttr *getDLLAttr(Decl *D) {
 
 /// Retrieve the depth and index of a template parameter.
-inline std::pair<unsigned, unsigned> getDepthAndIndex(NamedDecl *ND) {
+inline std::pair<unsigned, unsigned> getDepthAndIndex(const NamedDecl *ND) {
```
This could be an nfc change

antangelo replied:
I'll split this out into a separate patch

antangelo replied:
Done, see https://github.com/llvm/llvm-project/pull/113857

cor3ntin commented on clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -10616,6 +10616,43 @@ bool clang::isBetterOverloadCandidate(
+            FPT1->getNumParams() == FPT2->getNumParams()) {
+          const auto &P1 = FPT1->getParamTypes();
+          const auto &P2 = FPT2->getParamTypes();
+          bool ParamsHaveSameType = llvm::all_of(
+              llvm::zip(P1, P2),
```
We apparently have `all_of_zip`

antangelo replied:
I've updated this to use `all_of_zip`

cor3ntin commented on clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -10616,6 +10616,43 @@ bool clang::isBetterOverloadCandidate(
+          const auto &P2 = FPT2->getParamTypes();
+          bool ParamsHaveSameType = llvm::all_of(
+              llvm::zip(P1, P2),
+              [&](const std::tuple<const QualType &, const QualType &> &pair) {
```
```suggestion
              [&S](const std::tuple<const QualType &, const QualType &> &pair) {
```

antangelo replied:
Done

cor3ntin commented on clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -11758,6 +11795,40 @@ static void DiagnoseBadDeduction(Sema &S, NamedDecl *Found, Decl *Templated,
     }
 
+    // Errors in deduction guides from inherited constructors
+    // will present as substitution failures in the mapping
```
```suggestion
    // will manifest as substitution failures in the corresponding
```

(not 100% sure what you mean by mapping?)

antangelo replied:
I've updated the comment. `Mapping` here refers to the partial specialization that maps the return type of the original deduction guide back to the derived class

cor3ntin commented on clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -11758,6 +11795,40 @@ static void DiagnoseBadDeduction(Sema &S, NamedDecl *Found, Decl *Templated,
+      CXXDeductionGuideDecl *Source = DG->getSourceDeductionGuide();
+      assert(Source &&
+             "Inherited constructor deduction guides must have a source");
+      QualType DeducedRecordType(
+          cast<ClassTemplateDecl>(DG->getDeducedTemplate())
+              ->getTemplatedDecl()
+              ->getTypeForDecl(),
+          0);
+      QualType InheritedRecordType(
+          cast<ClassTemplateDecl>(Source->getDeducedTemplate())
+              ->getTemplatedDecl()
+              ->getTypeForDecl(),
+          0);
```
At least, can we get a lambda to get a type from a DG?

antangelo replied:
Done

cor3ntin commented on clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -11758,6 +11795,40 @@ static void DiagnoseBadDeduction(Sema &S, NamedDecl *Found, Decl *Templated,
+          << DeducedRecordType << InheritedRecordType << TemplateArgString;
+
+      CXXConstructorDecl *Ctor = DG->getCorrespondingConstructor();
+      if (Ctor && Ctor->getBeginLoc().isValid())
```
When is `Ctor->getBeginLoc()` not valid?

antangelo replied:
I don't recall why I added that initially, I believe there was an example that triggered this but it no longer seems to be necessary and I can't think of conditions that would trigger it. I have removed that check accordingly.

cor3ntin commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -899,7 +902,8 @@ Expr *buildIsDeducibleConstraint(Sema &SemaRef,
       TypeTrait::BTT_IsDeducible, IsDeducibleTypeTraitArgs,
-      AliasTemplate->getLocation(), /*Value*/ false);
+      AliasTemplate->getLocation(),
+      /*Value*/ false);
```
```suggestion
      /*Value=*/false);
```

antangelo replied:
Done

cor3ntin commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -931,12 +935,73 @@ getRHSTemplateDeclAndArgs(Sema &SemaRef, TypeAliasTemplateDecl *AliasTemplate) {
+};
+
+// Build the type for a deduction guide generated from an inherited constructor
+// [over.match.class.deduct]p1.10:
```
```suggestion
// C++23 [over.match.class.deduct]p1.10:
```

antangelo replied:
Done

cor3ntin commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -1107,8 +1172,32 @@ BuildDeductionGuideForTypeAlias(Sema &SemaRef,
+                       ->getReturnType();
+    }
+
+    // We omit the deducible constraint for inherited constructor deduction
+    // guides because they would take precedence over the derived class' own
+    // deduction guides due to [over.match.best.general]p2.5 and
+    // [temp.func.order]p6.4 If the alias were not deducible in this case, the
+    // deduction guide would already not be deducible due to the partial
+    // specialization `CC<>` failing substitution.
```
Did we decide this is correct? Should it not be a conjunction?

hokein replied:
Not yet, there's been no response from the CWG on the [issue](https://github.com/cplusplus/CWG/issues/607). The current implementation follows the suggestion outlined in that issue, and I think it's fine as is. However, we should probably reference this issue in a comment for clarity.

You can also check our related [discussion](https://github.com/llvm/llvm-project/pull/98788#discussion_r1686255752).

cor3ntin commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -1398,5 +1807,30 @@ void Sema::DeclareImplicitDeductionGuides(TemplateDecl *Template,
       ->setDeductionCandidateKind(DeductionCandidate::Copy);
 
+  CXXRecordDecl *TemplatedDecl = Pattern->getTemplatedDecl();
+  if (getLangOpts().CPlusPlus23 && TemplatedDecl->hasDefinition()) {
```
Is that something we should support in C++20 as an extension?

antangelo replied:
Given the uncertainty with open CWG issues and the potential breakage of code on earlier standards versions, I think it's less risky to keep it constrained to C++23 for now. It's probably worth revisiting in the future, though.

cor3ntin reviewed (commented):
I think that this looks generally good

A few comments
 - can you give a description  if what needs to be done in subsequent PRs? (maybe cxx_status.html should say "partial"
 - can you add a release notes
 - can you add PCH or module tests for the changes to ASTReader/ASTWriter?

hokein reviewed (approved):
Thanks! It looks good from my side. Please address the comments from the other reviewers and wait a bit before landing, in case they have additional feedback.

antangelo committed:
Address PR feedback

- Rename SourceKind to SourceDeductionGuideKind and add a clarifying
  comment
- Add modules test case for source deduction guide AST Reader/Writer
  changes
- Update ParamsHaveSameType check to use all_of_zip and restrict lambda
  capture
- Update comment to replace mapping terminology to better refer to the
  return type partial specialization
- Refactor deduction guide deduced template type to a lambda
- Remove unneeded check for invalid Ctor begin loc
- Update standards comments to refer to C++23
- Fix a bugprone argument convention comment

antangelo committed:
clang-format

antangelo committed:
Merge branch 'main' of github.com:llvm/llvm-project into p2582r1

antangelo committed:
Fix ASTImporterTest

antangelo committed:
Change cxx_status to partial, comment about dependent using-declarators

antangelo commented:
Thanks for the reviews!

>     * can you give a description  if what needs to be done in subsequent PRs? (maybe cxx_status.html should say "partial"
The remaining tasks are:
- Supporting dependent using-declarators such as `using Derived::Base::Base`. These are somewhat annoying to deal with, there isn't much of an existing mechanism (that I could find, at least) to look up the base type from the raw dependent identifier.
- Supporting deduction guides that are declared after the first lookup of deduction guides on the derived class. This is unimplemented due to a similar issue in alias template deduction that needs to be resolved first.

I have updated the cxx_status to partial, and added a note about the unsupported depdendent using-declarators. This aligns with the status for alias template deduction, which is partial due to the feature macro not being set. I presume the issue of deduction guides declared after the first lookup of deduction guides would be fixed in both features at the same time, either in this patch set if it is fixed in upstream first, or after this patch set as a followup.

>     * can you add a release notes
There should already be a clang release note, are there any other spots besides ReleaseNotes.rst and cxx_status.html that need updating?

>     * can you add PCH or module tests for the changes to ASTReader/ASTWriter?
I have added module tests that cover the feature.

antangelo committed:
Merge branch 'main' of github.com:llvm/llvm-project into p2582r1

hokein commented:
> Thanks for the reviews!
> 
> > ```
> > * can you give a description  if what needs to be done in subsequent PRs? (maybe cxx_status.html should say "partial"
> > ```
> 
> The remaining tasks are:
> 
> * Supporting dependent using-declarators such as `using Derived::Base::Base`. These are somewhat annoying to deal with, there isn't much of an existing mechanism (that I could find, at least) to look up the base type from the raw dependent identifier.

Although there are FIXMEs in the code, could you also file a GitHub issue for better tracking?

> * Supporting deduction guides that are declared after the first lookup of deduction guides on the derived class. This is unimplemented due to a similar issue in alias template deduction that needs to be resolved first.
> 
> I have updated the cxx_status to partial, and added a note about the unsupported depdendent using-declarators. This aligns with the status for alias template deduction, which is partial due to the feature macro not being set. I presume the issue of deduction guides declared after the first lookup of deduction guides would be fixed in both features at the same time, either in this patch set if it is fixed in upstream first, or after this patch set as a followup.

Yes, the missing-after-first-lookup issue is already tracked in #103016, and it also affects alias template deduction (it’s the only blocker remaining to mark alias template deduction as a completed feature).

We can address this as a follow-up, and I think we should resolve it before the next Clang 20 release.

cor3ntin commented on clang/include/clang/AST/DeclCXX.h:
```diff
@@ -2010,6 +2031,25 @@ class CXXDeductionGuideDecl : public FunctionDecl {
+  /// Get the deduction guide from which this deduction guide was generated,
+  /// if it was generated as part of alias template deduction or from an
+  /// inherited constructor.
+  CXXDeductionGuideDecl *getSourceDeductionGuide() const {
```
```suggestion
  const CXXDeductionGuideDecl *getSourceDeductionGuide() const {
```
Would that break anything?

AaronBallman replied:
Yeah, we should be returning a `const` pointer from a `const` member function; we can add a non-`const` member function overload to return a non-`const` pointer if we need one.

antangelo replied:
Fixed

cor3ntin commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -1203,10 +1333,291 @@ void DeclareImplicitDeductionGuidesForTypeAlias(
+      Context, SourceLocation(), SourceLocation(),
+      ArrayRef<NamedDecl *>(TParam), SourceLocation(), nullptr);
+
+  std::string CCTemplateDeclName =
+      (Twine("__ctad_CC_") + BaseTD->getName() + "_to_" + Template->getName() +
+       "_" + Twine(BaseIdx))
+          .str();
+  IdentifierInfo *CCTemplateII = &Context.Idents.get(CCTemplateDeclName);
+  CXXRecordDecl *CCTemplateRD = CXXRecordDecl::Create(
```
Do we actually need a name here (or could the record be anonymous)?

antangelo replied:
I initially tried it as anonymous, and it worked in normal use, but I experienced some ICE crashes while debugging/dumping types that included it. I am not sure which internal assumptions are broken by this, if any, given that anonymous templates are not supported by the language, which is why I ended up naming it.

I also imagine that looking this up by its name will be the most efficient way to retrieve the `CC` and alias types again when we implement handling for deduction guides declared after the first use (since we won't be recreating these AST nodes, and storing references to them anywhere else seems too wasteful).

cor3ntin commented on clang/test/SemaCXX/cxx23-ctad-inherited-ctors.cpp:
```diff
@@ -0,0 +1,316 @@
+                                                    // expected-note 2{{implicit deduction guide declared as }}
+
+    using ExplicitBase<T>::ExplicitBase; // expected-note {{candidate template ignored: couldn't infer template argument 'T'}} \
+                                         // expected-note {{implicit deduction guide declared as 'template <NoPointers<> T, typename V> ExplicitDerived(V) -> typename __ctad_CC_ExplicitBase_to_ExplicitDerived_0<ExplicitBase<T>>::type'}} \
+                                         // expected-note {{candidate template ignored: could not match 'ExplicitBase<T>' against 'const char *'}} \
+                                         // expected-note {{candidate template ignored: could not deduce template arguments for 'ExplicitDerived<T>' from 'ExplicitBase<T>' [with T = const char *]}} \
+                                         // expected-note {{implicit deduction guide declared as 'template <NoPointers<> T> ExplicitDerived(ExplicitBase<T>) -> typename __ctad_CC_ExplicitBase_to_ExplicitDerived_0<ExplicitBase<T>>::type'}}
```
This is not the best diagnostics.
Could we detect implicit deduction guides and
show something like 

> `<implicit deduction guide for X inherited from Y>`
> note : `Y` declared here

antangelo replied:
By `<implicit deduction guide for X inherited from Y>` do you mean to show the declaration for the original deduction guide on the base instead of the transformed guide on the derived class?

antangelo replied:
I have updated the diagnostics so that we print the original source deduction guide, following the source pointer, and for each source pointer we follow I also now print the location that the relevant constructor is inherited.

AaronBallman commented on clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -10615,6 +10615,40 @@ bool clang::isBetterOverloadCandidate(
+
+        // ... and for each explicit function argument, the parameters of F1 and
+        // F2 are either both ellipses or have the same type
+        if (FPT1->isVariadic() == FPT2->isVariadic() &&
+            FPT1->getNumParams() == FPT2->getNumParams()) {
```
This seems wrong compared to the comment -- the comment says we should be looking at each explicit function argument, but this predicate only handles the case where both functions are variadic and have the same number of params.

Am I missing something?

erichkeane replied:
Is there a reason we can't just check the types of the functions here instead?  I would assume their prototypes to compare equal.  That said, the wording of `parameters are both ellipses or have the same type` seems intentionally designed for the purpose of excluding the return type (or, perhaps reference/cv specifiers?)?  

If so, I think I'd rather a variant of `hasSameType` that for something like, `hasSameParameterList`.  At that point, we could have IT check for similar variadics as well. We probably want to make sure that `(int, ...)` works too.

Also, I don't think the initial check is right?  Shouldn't that be an 'or' there instead of '&&'?  If both are `NumParams == 0 && isVariadic`, then we want to hit this condition as well.

antangelo replied:
The return types will not be equal as the inherited deduction guides will have some `typename CC<R>::type` as the return type, while non-inherited deduction guides will just have a template specialization type for the class template.

If both functions are variadic and have `getNumParams() == 0`, then that boolean expression evaluates to true right? Can you elaborate on what you think the behavior there should be? I may be misunderstanding what the standards wording on this part is asking.

cor3ntin replied:
We could assert that `getNumNonObjectParams` is `false` (I don't think we can get explicit objects anywhere, right? Otherwise, the code looks correct, but I agree that something like a "haveSameNonObjectParameters" function would improve readability

AaronBallman commented on clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -11757,6 +11791,42 @@ static void DiagnoseBadDeduction(Sema &S, NamedDecl *Found, Decl *Templated,
+    // will manifest as substitution failures in the return type
+    // partial specialization, so we show a generic diagnostic
+    // in this case.
+    if (auto *DG = dyn_cast<CXXDeductionGuideDecl>(Templated);
```
```suggestion
    if (const auto *DG = dyn_cast<CXXDeductionGuideDecl>(Templated);
```
and handle const correctness further down as well.

antangelo replied:
Fixed

AaronBallman commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -931,12 +935,73 @@ getRHSTemplateDeclAndArgs(Sema &SemaRef, TypeAliasTemplateDecl *AliasTemplate) {
+TypeSourceInfo *buildInheritedConstructorDeductionGuideType(
+    Sema &SemaRef, const InheritedConstructorDeductionInfo &Info,
+    TypeSourceInfo *SourceGuideTSI) {
+  auto &Context = SemaRef.Context;
```
```suggestion
  ASTContext &Context = SemaRef.Context;
```

antangelo replied:
Fixed

AaronBallman commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -931,12 +935,73 @@ getRHSTemplateDeclAndArgs(Sema &SemaRef, TypeAliasTemplateDecl *AliasTemplate) {
+  NewTL.setRParenLoc(TL.getRParenLoc());
+  NewTL.setExceptionSpecRange(TL.getExceptionSpecRange());
+  NewTL.setLocalRangeEnd(TL.getLocalRangeEnd());
+  for (unsigned I = 0, E = NewTL.getNumParams(); I != E; ++I)
```
Do we have to do something special here to handle a variadic?

antangelo replied:
I don't believe so, from the comment on `FunctionProtoType` it seems the source location for ellipses is stored in the type itself, and there isn't any mention of them in `FunctionProtoTypeLoc`
https://github.com/llvm/llvm-project/blob/476b208e0115e766605e9f850982996a1d51c287/clang/include/clang/AST/Type.h#L5111-L5112

I have added a C-style variadic test case as well to verify that they work as expected.

AaronBallman commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -1203,10 +1333,291 @@ void DeclareImplicitDeductionGuidesForTypeAlias(
 }
 
+// Check if a template is deducible as per [dcl.type.simple]p3
+static bool IsDeducibleTemplate(TemplateDecl *TD) {
```
```suggestion
static bool IsDeducibleTemplate(const TemplateDecl *TD) {
```
and threaded through the rest of the function?

antangelo replied:
Fixed

AaronBallman commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -1203,10 +1333,291 @@ void DeclareImplicitDeductionGuidesForTypeAlias(
+      CloneTDI.SubstTemplateParams(AliasTPL);
+  CloneScope.Exit();
+  for (NamedDecl *Param : *PartialSpecTPL) {
+    if (auto *TTP = dyn_cast<TemplateTypeParmDecl>(Param)) {
+      TTP->removeDefaultArgument();
+    } else if (auto *NTTP = dyn_cast<NonTypeTemplateParmDecl>(Param)) {
+      NTTP->removeDefaultArgument();
+    } else if (auto *TTP = dyn_cast<TemplateTemplateParmDecl>(Param)) {
+      TTP->removeDefaultArgument();
+    }
```
```suggestion
    if (auto *TTP = dyn_cast<TemplateTypeParmDecl>(Param))
      TTP->removeDefaultArgument();
    else if (auto *NTTP = dyn_cast<NonTypeTemplateParmDecl>(Param))
      NTTP->removeDefaultArgument();
    else if (auto *TTP = dyn_cast<TemplateTemplateParmDecl>(Param))
      TTP->removeDefaultArgument();
```

antangelo replied:
Fixed

mizvekov commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -931,12 +935,73 @@ getRHSTemplateDeclAndArgs(Sema &SemaRef, TypeAliasTemplateDecl *AliasTemplate) {
+  const auto *FPT = SourceGuideTSI->getType()->getAs<FunctionProtoType>();
+  assert(FPT && "Source Guide type should be a FunctionProtoType");
+
+  // This substitution can fail in cases where the source return type
+  // is not dependent and the derived class is not deducible
+  Sema::SFINAETrap Trap(SemaRef);
```
This is not clear to me, why would the substitution fail if the source is not dependent? Would there instead be simply no substitution? Also why is the derived class not being deducible relevant here, since we are not performing any deductions?

antangelo replied:
This code is substituting the return type `R` of the base deduction guide into the CC template type `typename CC<R>::type` to form the return type for the inherited deduction guide. When `R` is dependent, the resulting type is also dependent, but if `R` is not dependent this will cause the CC template to be instantiated here.

The motivating test case for when this is problematic is:
```c++
template <typename T>
struct NonTemplateDGuideBase {
    NonTemplateDGuideBase(T);
};

NonTemplateDGuideBase(int) -> NonTemplateDGuideBase<int>;
NonTemplateDGuideBase(const char *) -> NonTemplateDGuideBase<const char *>;

template <typename T>
concept NoPointers = !requires(T t) { *t; };

template <NoPointers T>
struct NonTemplateDGuideDerived : public NonTemplateDGuideBase<T> {
    using NonTemplateDGuideBase<T>::NonTemplateDGuideBase;
};

NonTemplateDGuideDerived ntdg_char("");
```

Since constraints are involved, substituting a non-dependent `R` into the CC type causes constraints to be evaluated, and if the derived class is not deducible given base type `R` (e.g. when the constraints evaluate to false), this substitution will fail. This is not an issue if `R` is dependent as the constraints will evaluate later during argument deduction, or if the derived class is deducible given `R` as `CC<R>` will be a valid substitution (e.g. in the above example, the `int` deduction guide does not cause issues).

mizvekov replied:
I see. But why then are we suppressing the diagnostics in this case?
If there is a substitution failure otherwise, is that currently being diagnosed somewhere else?

antangelo replied:
If we don't suppress diagnostics here, then these substitution failure errors would appear at the time that implicit deduction guides are declared (regardless of whether or not the failing deduction guide was actually used).

Right now any deduction guide that fails substitution here will be skipped, and not inherited by the base. Unfortunately this means that we don't get diagnostics for them, but there isn't a great way to propagate the error through to the point where deduction of the derived class fails. We don't have the return type to create a valid deduction guide AST node, like we do in other cases where we know deduction will fail up front (e.g. with default parameters not appearing in the base), so some other mechanism is needed to preserve this information until deduction of the derived class has failed and the candidate notes are being printed.

mizvekov replied:
I see, that is a bummer. This is worth pointing out in a FIXME.
Are you planning on working on a fix for that?

antangelo replied:
I have added a FIXME, I can work on a fix for it in a followup but it will likely need more invasive changes to the AST in order to preserve the necessary information.

mizvekov commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -1173,25 +1276,52 @@ void DeclareImplicitDeductionGuidesForTypeAlias(
-      auto *Transformed = cast<FunctionDecl>(buildDeductionGuide(
-          SemaRef, AliasTemplate, /*TemplateParams=*/nullptr,
+
+      QualType ReturnType =
+          cast<FunctionProtoType>(FunctionType->getType())->getReturnType();
+      if (FromInheritedCtor) {
+        FunctionType = buildInheritedConstructorDeductionGuideType(
+            SemaRef, *FromInheritedCtor, FunctionType);
+        if (!FunctionType)
+          continue;
+        ReturnType = FunctionType->getType()
+                         .getTypePtr()
+                         ->getAs<FunctionProtoType>()
+                         ->getReturnType();
+      }
```
There is a lot of commonality with this block here and another that appears in `BuildDeductionGuideForTypeAlias`, can we factor it out?

antangelo replied:
I have moved the `ReturnType` logic into `buildInheritedConstructorDeductionGuideType` and now return a pair from that function.

antangelo committed:
Address PR feedback and fix crash when declaring inherited/alias DG from an inherited DG

PR feedback addressed:

- Const correctness in CXXDeductionGuideDecl ASTNode and throughout
  added code
- Add FIXME for not emitting diagnostic in case where CC type
  substitution fails
- When emitting notes for deduction guide declarations, follow the
  source pointer for inherited deduction guides so the first source
  guide is shown instead of the inherited guide, and emit a note showing
  the point(s) that the constructors are inherited
- Refactor buildInheritedConstructorDeductionGuideType to return the
  ReturnType directly
- Add test for C-style variadic constructor
- Replace some uses of 'auto' with the full type
- Update if/else-if formatting

antangelo committed:
clang-format

antangelo committed:
Merge remote-tracking branch 'upstream/main' into p2582r1

antangelo committed:
Pass context into getInjectedTemplateArgs

antangelo commented on clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -11757,6 +11791,42 @@ static void DiagnoseBadDeduction(Sema &S, NamedDecl *Found, Decl *Templated,
+            diag::
+                note_ovl_candidate_inherited_constructor_deduction_failure_source)
+            << InheritedRecordType;
+      return;
```
I'm using the source location of the inherited constructor deduction guide to store the location of the using decl that inherits the base constructors (as it is otherwise difficult to pass this information around), so the notes for deduction guide failure will point there instead of at the source constructor (or deduction guide in the explicit case).

I have added additional diagnostics that print the source deduction guide declaration (with the source deduction guide's location) for both implicit and explicit guides when we note the deduction guide declarations for implicit guides as a workaround, which also removes the need for the above note by extension.

antangelo commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -942,34 +946,145 @@ getRHSTemplateDeclAndArgs(Sema &SemaRef, TypeAliasTemplateDecl *AliasTemplate) {
-    FReturnType = ET->getNamedType()->getAs<TemplateSpecializationType>();
-  assert(FReturnType && "expected to see a return type");
+  ArrayRef<TemplateArgument> FTemplateArgs =
+      getTemplateArgumentsFromDeductionGuideReturnType(
```
@hokein I would like to re-request your review for these new changes to `BuildDeductionGuideForTypeAlias` to handle declaring alias and inherited deduction guides from an inherited deduction guide as the source (e.g. in cases with deeper class hierarchies or aliased class templates that inherit deduction guides).

erichkeane commented on clang/include/clang/AST/DeclCXX.h:
```diff
@@ -1960,24 +1960,44 @@ class ExplicitSpecifier {
+  // Represents the relationship between this deduction guide and the
+  // deduction guide that it was generated from (or lack thereof).
+  // See the SourceDeductionGuide member for more details.
+  enum class SourceDeductionGuideKind : unsigned char {
```
```suggestion
  enum class SourceDeductionGuideKind : uint8_t {
```

I thought I saw elsewhere we do this reasonably consistently?  Else, feel free to ignore.

erichkeane commented on clang/include/clang/AST/DeclCXX.h:
```diff
@@ -2010,6 +2032,25 @@ class CXXDeductionGuideDecl : public FunctionDecl {
+  }
+
+  void setSourceDeductionGuide(CXXDeductionGuideDecl *DG) {
+    SourceDeductionGuide.setPointer(DG);
```
Do we want to assert for `!DG`?

antangelo replied:
I opted not to assert here and in `setSourceDeductionGuideKind` because it makes the `ASTReader` changes more readable (which otherwise would have to do a check on at least one of these fields before setting it on the incoming `CXXDeductionGuideDecl` node, but maybe that's preferable).

erichkeane commented on clang/include/clang/AST/DeclCXX.h:
```diff
@@ -2010,6 +2032,25 @@ class CXXDeductionGuideDecl : public FunctionDecl {
+  }
+
+  void setSourceDeductionGuideKind(SourceDeductionGuideKind SK) {
+    SourceDeductionGuide.setInt(SK);
```
PERHAPS here too?  Assert `SK == None`?

erichkeane commented on clang/lib/AST/DeclCXX.cpp:
```diff
@@ -2246,18 +2246,21 @@ CXXDeductionGuideDecl *CXXDeductionGuideDecl::Create(
 
 CXXDeductionGuideDecl *
 CXXDeductionGuideDecl::CreateDeserialized(ASTContext &C, GlobalDeclID ID) {
   return new (C, ID) CXXDeductionGuideDecl(
```
For `Expr`s we usually have a `CreateEmpty` that has an `empty`/`empty-ish` private constructor associated with it. or something like that.... this is very much making me wish we had the same for this one.

erichkeane commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -942,34 +946,145 @@ getRHSTemplateDeclAndArgs(Sema &SemaRef, TypeAliasTemplateDecl *AliasTemplate) {
 
+struct InheritedConstructorDeductionInfo {
+  // Class template for which we are declaring deduction guides
+  // This is `C` in the standard wording
```
Comments are supposed to end in full-stops, so I think there is a sprinkling of `.` required here/elsewhere here (other than standards quotes).

erichkeane commented on clang/lib/Sema/SemaTemplateDeductionGuide.cpp:
```diff
@@ -942,34 +946,145 @@ getRHSTemplateDeclAndArgs(Sema &SemaRef, TypeAliasTemplateDecl *AliasTemplate) {
+  Args.addOuterTemplateArguments(Info.DerivedClassTemplate,
+                                 TemplateArgument(FPT->getReturnType()),
+                                 /*Final=*/false);
+  Args.addOuterRetainedLevels(Info.DerivedClassTemplate->getTemplateDepth());
```
As we sure this is sufficient and shouldn't be calling the `getInstantiationArgs`?  It might need help with Deduction guides, but is meant to get the full depth of the deduction guide.  But `OuterRetainedLevels` means it won't fully instantiate (and those levels will just not fill anything in).  Though is that perhaps what is intended here?

erichkeane reviewed (commented):
I don't think I'm needed on this review again, but did a pass through and had some comments/suggestions.  Nothing binding, so I'll let others discuss this/accept this.

cor3ntin commented:
@antangelo @hokein What's the status here?
If this is not merged by next week it won't make clang 20 (and there are unaddressed comments)

Thanks!

antangelo commented:
I'm still working on addressing the remaining comments, primarily the one from Erich about insufficient handling of nested templates. There are some issues when the derived class is in a nested template, and the base specifier depends on the outer template arguments, where the CC type we create has arguments substituted at the wrong levels when used in a deduction guide. I haven't had time to debug this fully, but my suspicion is that it has something to do with placing the CC template inside the outer template specialization without it being defined in the outer template pattern (but this suspicion may be wrong).

I don't know if I'll be able to fix these issues before clang 20 branches. The best way forward might be for me to:

1. Split out the deduction guide source tracking into its own patch, so that completing the alias deduction guide feature is no longer blocked on this (which is something I probably should've done to begin with)
2. If we want to land the portion of this feature that is done so far, we can land it without support for nested templates where the base specifier depends on outer template arguments (with some detection for this case and perhaps a diagnostic note) and then revisit this in a follow-up. I'll defer to other maintainers as to whether or not this is an option we want to pursue.
3. Otherwise, this feature will likely need to wait for clang 21.

cor3ntin commented:
Both 1 and 2 seems like good options @hokein @erichkeane 
(especially 1 is probably something we should try to do)

erichkeane commented:
> Both 1 and 2 seems like good options @hokein @erichkeane (especially 1 is probably something we should try to do)

I like 1, but not really 2, I think we should wait for this to be reasonably complete before merging it, particularly this close to the release.  I'd probably be ok with 2 if we wait a few weeks until after the release, so that we can deal with the fallout with less stress, and fix it before Clang21 release.

antangelo commented:
I have split the source deduction guide tracking into #123875, and will continue working on fixing the remaining issues here to land P2582R1 for clang 21.

hokein commented:
> I like 1, but not really 2, I think we should wait for this to be reasonably complete before merging it, particularly this close to the release. I'd probably be ok with 2 if we wait a few weeks until after the release, so that we can deal with the fallout with less stress, and fix it before Clang21 release.

+1. Rushing a new feature just before the release increases the risk of issues and adds unnecessary stress.

I believe we should find a way to land this patch, even if it’s not fully complete (it has been pending for months). Landing it after the release cut seems like a good approach. This would allow for incremental development while giving us more time to address any fallout before clang 21.

cor3ntin commented:
@antangelo @erichkeane @hokein Should we try to land this now?

antangelo commented:
This patch isn't ready to land yet, unfortunately I haven't been able to find much time to work on it recently. The main remaining issue is that deduction fails for nested templates where the base specifier depends on an outer template argument, and I have been unable to find the root cause for that yet. I anticipate this may cause existing code to break if landed as-is, so I think this patch should wait until it the issue can be fixed.
